### ISSUE TITLE
Duplicate primary key while creating instances of Django model

### ISSUE THREAD

**Issue by timmartin** [HypothesisWorks/hypothesis#1307]
I'm trying to create a strategy for testing a Django model, and I'm repeatedly getting integrity errors for a duplicate primary key.

I believe the problem is in the way that Django objects are constructed using `get_or_create` when the primary key is one of the values generated as part of the strategy. We shouldn't pass all the fields to `get_or_create`, we should pass the primary key as a keyword arg and the remaining values as the `defaults` keyword arg. This way only the primary key is used for determining duplicate rows, rather than requiring the values of all fields to match.

This problem hit because my model has both a `OneToOneField` with `primary_key=True` and an `AutoLastModifiedField`. What I think happens is:

- The instance is created and written to DB. The `AutoLastModifiedField` updates the value of `modified` at the last minute.
- Hypothesis creates the same values and tries to create the object with `get_or_create`. However, this currently has to match every field. The existing row doesn't match because the `modified` value has changed, so a new object is created.
- The new object can't be created, because there is a clashing primary key.

```python
    class User(AbstractBaseUser):
        uuid = models.UUIDField(
            # ...
        )

        first_name = models.CharField(max_length=64)
        last_name = models.CharField(max_length=64)

    class Teacher(models.Model):
        user = models.OneToOneField(
            User,
            on_delete=models.CASCADE,
            primary_key=True
        )

        modified = AutoLastModifiedField('modified')
```

**Comment by timmartin** [comment]
I have [a branch that fixes this](https://github.com/timmartin/hypothesis/tree/django-primary-key), though I haven't fully tested / code reviewed it yet.

**Comment by timmartin** [comment]
Thanks for the feedback. My intention was definitely to clean this up and get it ready for PR.

I agree that there's a broader question about whether `update_or_create` does the right thing, though I don't think there's any other local solution. If the new object duplicates an existing primary key then it's the same object as the other one, by definition. But I do wonder whether the right solution is for the strategy to avoid calling the create method twice with the same args, (except when randomly striking the same values by sheer chance). The idiomatic Django thing would be to call `.create()` when you know the row shouldn't already exist, and `.save()` when you know the object already exists.

I still haven't got to the bottom of why the code is attempting to create the same object twice. I reliably hit this problem when just doing a single `.example()` call, so it's not (apparently) the case that it's randomly drawing the same value twice from a large space. My guess is that the code just relies on the fact that drawing objects ought to be idempotent, and it might be a pain to modify this assumption to work with Django. If this is the case then my patch is probably as close as we can get to a "make this Django model object exist" idempotent function.

For now I'll clean up and UT the code I already have and make a PR.

**Comment by timmartin** [comment]
On the specifics of the patch: I wasn't very happy with my `create_object` closure and I'd like to get rid of it, but I was worried that the new `constructor_args` structure nests the strategies deeper in the arguments, and that `strategies.builds()` isn't going to do the right thing. Previously we were returning a dict of strategies, now we're returning (partialy) a dict of dict of strategies, and `strategies.builds()` isn't going to do the right thing with that, I think. That's why I tried to do it in the closure, so `builds()` can do its work with a straight collection of strategies and we can build the Django constructor args after the strategies have yielded real values.

I was more inclined to move the Django constructor_args code into `models_impl`, since it seems like it belongs there. But then we have to deal with passing the primary key name into the function, which my closure was the lazy solution to.

I'll think about it a bit more, but comments are welcome.

**Comment by timmartin** [comment]
I had a little trouble figuring out a sensible unit test, but in the process figured out a lot more about what's going wrong here. I tried to get a more basic unit test than just calling `example()`, and I experimented with a couple of cases of trying to draw twice from the same strategy and hit the problem. I wasn't able to hit the problem this way, hence my UT just consisting of calling `example()` and checking that it isn't rejected.

The cause of the problem turns out to be that in `core.find()` we [call runner.run(), which draws from the strategy](https://github.com/timmartin/hypothesis/blob/master/hypothesis-python/src/hypothesis/core.py#L1153), then later on [construct a copy of the buffer and re-draw the same object again](https://github.com/timmartin/hypothesis/blob/master/hypothesis-python/src/hypothesis/core.py#L1160).

In theory I could reproduce this in the UT by cloning a buffer and then drawing using the original buffer and the cloned buffer, but I couldn't figure out how to do that cleanly without messing too much with the internals.
